gorm/transaction.go

```diff
@@ -57,15 +81,29 @@ func (t *Transaction) AddAfterCommitHook(hooks ...func(context.Context)) {
 	t.afterCommitHook = append(t.afterCommitHook, hooks...)
 }
 
-// BeginFromContext will extract transaction wrapper from context and start new transaction.
+// GetReadOnlyDBInstance returns the read only database instance stored in the ctx
+func GetReadOnlyDBInstance(ctx context.Context) (*gorm.DB, error) {
+	dbRO, ok := ctx.Value(roDBKey).(*gorm.DB)
+	if !ok {
+		return nil, ErrNoReadOnlyDB
+	}
+	return dbRO, nil
+}
+
+// BeginFromContext will extract transaction wrapper from context and start new transaction if transaction is not set to read only otherwise it will return read only database instance
 // As result new instance of `*gorm.DB` will be returned.
 // Error will be returned in case either transaction or db connection info is missing in context.
 // Gorm specific error can be checked by `*gorm.DB.Error`.
 func BeginFromContext(ctx context.Context) (*gorm.DB, error) {
+	logger := ctxlogrus.Extract(ctx)
 	txn, ok := FromContext(ctx)
 	if !ok {
 		return nil, ErrCtxTxnMissing
 	}
+	if txn.readOnly == true {
```

Comment:
We'll need to update the `BeginFromContext` tests for this change. The existing tests do not check that a transaction from the read-write parent is returned in the current cases, and we'll have to add a test to confirm the read-only database reference is returned on a read-only context.